### PR TITLE
Revise supported databases and requirements/recommendations docs

### PR DESCRIPTION
**Pull request by josh-wong**
## Description

This PR revises the document listing the databases that ScalarDB supports and the document explaining the requirements and recommendations for specific types of databases.

## Related issues and/or PRs

- https://github.com/scalar-labs/scalardb/pull/1012

## Changes made

- In `scalardb-supported-databases.md`, I separated the tables to that they are easier to skim. As we continue to add databases and versions, having tables separated by product will be easier to read. (Previously, the table required scrolling horizontally, which some users may not know is possible and would eventually become pretty long.)
- In `requirements.md`, I revised the text primarily from a grammatical point of view.

## Branches this PR applies to

- `master`
- `3.10`
- `3.9`
- `3.8`
- `3.7`
- `3.6`
- `3.5`

## Testing done

I ran our docs site locally and confirmed that the two pages appeared as expected. The following screenshot shows how the `scalardb-supported-databases.md` would look. (Apologies for the long screenshot😅)

PR-1020-screenshot

## Checklist

- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] The documentation has been updated to reflect the changes.
- [x] Any remaining open issues linked to this PR are documented and up-to-date (Jira, GitHub, etc.).
- [x] Tests (unit, integration, etc.) have been added for the changes.
- [x] My changes generate no new warnings.
- [x] Any dependent changes in other PRs have been merged and published.

## Additional notes (optional)

Please see my comments in the docs.

### CODE REVIEW

File: docs/scalardb-supported-databases.md
Comment (josh-wong) [inline docs/scalardb-supported-databases.md:2]: Are there any additional databases and/or versions that we need to add?

File: docs/scalardb-supported-databases.md
Comment (josh-wong) [inline docs/scalardb-supported-databases.md:27]: I couldn't find any previous PRs or discussions about ScalarDB supporting Cassandra 4.0 or 4.1. Are those versions not officially supported? If not, I think we should add an ❌ to those versions in this table. 

Or, if we know that those versions work with ScalarDB but we haven't officially tested them, we can mention "Not tested". 

What do you think about that?

File: docs/scalardb-supported-databases.md
Comment (komamitsu) [reply docs/scalardb-supported-databases.md:2]: I saw we discussed MariaDB https://github.com/scalar-labs/scalardb/pull/1012#discussion_r1300823307 although I'm not sure which ScalarDB versions are supposed to support them.

File: docs/scalardb-supported-databases.md
Comment (komamitsu) [reply docs/scalardb-supported-databases.md:27]: (Posted [a related comment](https://github.com/scalar-labs/scalardb/pull/1012/files#r1302364828))

**Review by komamitsu — APPROVED** [review]
There are a few open discussions, but overall LGTM! 👍

File: docs/scalardb-supported-databases.md
Comment (brfrn169) [reply docs/scalardb-supported-databases.md:27]: Yeah, currently, we don't support Cassandra 4.0 and 4.1.

File: docs/scalardb-supported-databases.md
Comment (brfrn169) [inline docs/scalardb-supported-databases.md:35]: I see a broken table for this:
https://github.com/scalar-labs/scalardb/blob/fca2294971edcfda26868188798a03ac1623e990/docs/scalardb-supported-databases.md#apache-cassandra

Can we fix it?

**Review by brfrn169 — COMMENTED** [review]
Left several comments. Please take a look when you have time!

File: docs/index.md
Comment (josh-wong) [reply docs/index.md:38]: Good catch! Thanks!

File: docs/scalardb-supported-databases.md
Comment (josh-wong) [reply docs/scalardb-supported-databases.md:27]: Got it. To reduce potential support inquiries, I've marked Cassandra 4.0 and 4.1 as unsupported.

File: docs/scalardb-supported-databases.md
Comment (josh-wong) [reply docs/scalardb-supported-databases.md]: Ah yes, I had added columns for Cassandra 4.0 and 4.1 and forgot to add extra extra bars after the header row. Show be fixed now.

Interesting that the docs site rendered the tables but GitHub didn't.🤔

File: docs/scalardb-supported-databases.md
Comment (josh-wong) [reply docs/scalardb-supported-databases.md:2]: Thanks! I'll keep an eye on that PR so that this doc can be consistent with that compatibility check.

Comment (brfrn169) [reply docs/scalardb-supported-databases.md:2]: @Torch3333 is working on database-compatibility testing now. And I think he can update this document after finishing his work. So we can proceed with the current databases and versions in this PR. Thanks.

**Review by Torch3333 — APPROVED** [review]
LGTM, thank you!

File: docs/scalardb-supported-databases.md
Comment (josh-wong) [inline docs/scalardb-supported-databases.md:121]: @brfrn169 Based on the comment in https://github.com/scalar-labs/scalardb/pull/1012#discussion_r1302373516, I think we should add PostreSQL 15 to this table. 

Do ScalarDB versions 3.4 to 3.10 all support PostreSQL 15?

File: docs/scalardb-supported-databases.md
Comment (josh-wong) [inline docs/scalardb-supported-databases.md:73]: @brfrn169 Based on the comment in https://github.com/scalar-labs/scalardb/pull/1012#discussion_r1302373516, I think we should add Amazon Aurora PostreSQL 15 to this table. 

Do ScalarDB versions 3.4 to 3.10 all support Amazon Aurora PostreSQL 15?

File: docs/scalardb-supported-databases.md
Comment (josh-wong) [inline docs/scalardb-supported-databases.md:109]: @brfrn169 Based on the comment in https://github.com/scalar-labs/scalardb/pull/1012#discussion_r1302374704, I think we should add Oracle 23c to this table. 

Do ScalarDB versions 3.4 to 3.10 all support Oracle 23c?

File: docs/scalardb-supported-databases.md
Comment (brfrn169) [reply docs/scalardb-supported-databases.md:121]: @josh-wong 

@Torch3333 is working on the database-compatibility testing to check that. Do you want to wait for it before merging this PR? Or do you want to merge it with with the current databases and versions and update it after finishing the test?

Comment (josh-wong) [reply docs/scalardb-supported-databases.md:121]: @brfrn169 Ah, I see. I assumed, since the PR was merged, that those PostgreSQL and Oracle 23c were confirmed to be compatible.

No need to wait for database-compatibility testing to be finished. We can proceed with merging this PR and update later as needed👍

File: docs/scalardb-supported-databases.md
Comment (josh-wong) [reply docs/scalardb-supported-databases.md:73]: **Note:** See https://github.com/scalar-labs/scalardb/pull/1020#discussion_r1303972487

File: docs/scalardb-supported-databases.md
Comment (josh-wong) [reply docs/scalardb-supported-databases.md:109]: **Note:** See https://github.com/scalar-labs/scalardb/pull/1020#discussion_r1303972487

File: docs/requirements.md
Comment (feeblefakie) [inline docs/requirements.md:29]:
```suggestion
If you must perform operations or transactions on read replicas, you can configure your application to perform read-write-mixed transactions against a single master instance and read-only transactions against read replicas. The resulting transaction schedules with the configuration would not be strict serializable (linearizable and serializable) but would be serializable.
```
`schedule` here means (part of) the history of transactions, so it cannot be replaced with `configuration`.

File: docs/requirements.md
Comment (feeblefakie) [inline docs/requirements.md:5]: It's just a heads-up.
We also need a requirements section for CosmosDB since we need to set `strong` for consistency?

File: docs/scalardb-supported-databases.md
Comment (feeblefakie) [inline docs/scalardb-supported-databases.md:27]: Can we omit `Apache` in each column? I feel it's a bit redundant and makes the table big.
Ditto for the other tables. (omitting `Amazon` and `Azure`?)

**Review by feeblefakie — COMMENTED** [review]
Left some comments and suggestions. PTAL!

File: docs/requirements.md
Comment (brfrn169) [reply docs/requirements.md:5]: Yeah, that's correct. Thank you for the heads-up!

File: docs/scalardb-supported-databases.md
Comment (josh-wong) [reply docs/scalardb-supported-databases.md:27]: I'm OK with that since we already spell out the name of the products in the headings above each table.

File: docs/requirements.md
Comment (josh-wong) [inline docs/requirements.md:25]: @brfrn169 I added the following according to @feeblefakie's comment in https://github.com/scalar-labs/scalardb/pull/1020/files#r1303994693.

Is there any more information that we can add here, such as why **Strong** needs to be configured?

File: docs/requirements.md
Comment (josh-wong) [reply docs/requirements.md:5]: I've added a section about this to lines [21-25](https://github.com/scalar-labs/scalardb/pull/1020/files#diff-71c6b87612afd48ee2b295f885d377d87f291385b3b5ff796e71947739393994R21-R25). PTAL!

File: docs/requirements.md
Comment (brfrn169) [reply docs/requirements.md:25]: We require **Strong** consistency in CosmosDB because the ScalarDB transaction protocol (Consensus Commit) needs linearizable reads. The **Strong** consistency level provides this linearizability guarantee:
https://learn.microsoft.com/en-us/azure/cosmos-db/consistency-levels#strong-consistency

Comment (josh-wong) [reply docs/requirements.md]: @brfrn169 Thanks for the extra details! I've added a couple extra sentences to explain why this setting is necessary.

File: docs/requirements.md
Comment (kota2and3kan) [inline docs/requirements.md:37]:
```suggestion
Also, from the perspective of the specification of ScalarDB, the write operation sometimes occurs against the underlying database to update metadata for transaction management even if the transaction is read-only. Therefore, you cannot use read replica under ScalarDB. Note that you can still use a read replica as a backup and standby even if you follow this recommendation.
```
Before we discussed `whether ScalarDB supports read replica or not`. And, if my remembering is correct, as a result of the discussion, we decided that we don't support read replicas.

Note: You can see the result in the following issue in Jira.
https://scalar-labs.atlassian.net/browse/DLT-13744

So, I left a suggestion to apply the above things to this document.

@josh-wong 
Could you please take a look at my suggestion from a grammatical point of view?

@feeblefakie @brfrn169 
Could you please take a look at my suggestion from a technical point of view?

**Review by kota2and3kan — COMMENTED** [review]
Thank you for the update!
The new separated tables look good to me!
I left two suggestions for the `Requirements and Recommendations` document.
Please take a look when you have time!

File: docs/requirements.md
Comment (josh-wong) [reply docs/requirements.md:7]: Good catch! I'm sad the spell check feature in the Markdown linting tool I have installed didn't catch that...

File: docs/requirements.md
Comment (brfrn169) [reply docs/requirements.md:37]: @kota2and3kan You are right. We can't use read replicas when using `EXTRA_WRITE`.

Comment (kota2and3kan) [reply docs/requirements.md:37]: @brfrn169 
> You are right. We can't use read replicas when using EXTRA_WRITE. Thanks.

Thank you for your confirmation! I will update this sentence to make it simpler (We only state `read replica is not supported). 

So, in this document, I don't mention concrete things like `EXTRA_WRITE` since it might lead to confusion.

@feeblefakie  
> For how to correct it,
> how about simply saying something like this?
> We can put the reason as a side note or something, but I think we should not start with reasoning.

Ah, I see. Indeed, describing too concrete internal information may cause confusion for users.

I agree with your suggestion that describe it simply. I will update the sentence that I suggested.

Comment (kota2and3kan) [reply docs/requirements.md:37]: @feeblefakie @brfrn169 @josh-wong 
```suggestion
With this in mind, you must perform all operations or transactions against a single master instance so that you can achieve linearizability and avoid worrying about consistency issues in your application. In other words, ScalarDB does not support read replicas. Note that you can still use a read replica as a backup and standby even if you follow this guideline.
```

I updated my suggestion!

* Remove concrete (internal) reasons.
* Add `ScalarDB does not support read replicas` explicitly to make requirements clear and avoid unnecessary inquiries from users.
* Explain `You can still use a read replica as a backup and standby` as a note.

Please take a look again when you have time!

Comment (feeblefakie) [reply docs/requirements.md:37]: @kota2and3kan Thank you! It looks good.
@josh-wong Please review this from an English perspective.

**Review by feeblefakie — APPROVED** [review]
LGTM! Thank you!

@josh-wong Please review the suggested contents from Takanari.

File: docs/requirements.md
Comment (josh-wong) [reply docs/requirements.md:37]: LGTM! The only change I made was to put the "Note that ..." part into a notice block.

```suggestion
With this in mind, you must perform all operations or transactions against a single master instance so that you can achieve linearizability and avoid worrying about consistency issues in your application. In other words, ScalarDB does not support read replicas. 

{% capture notice--info %}
**Note**

You can still use a read replica as a backup and standby even when following this guideline.
{% endcapture %}

<div class="notice--info">{{ notice--info | markdownify }}</div>
```

Comment (kota2and3kan) [reply docs/requirements.md]: > The only change I made was to put the "Note that ..." part into a notice block, which I should have done earlier.

Looks good! Thank you!

**Review by kota2and3kan — APPROVED** [review]
LGTM! Thank you!

**Review by brfrn169 — APPROVED** [review]
LGTM! Thank you!

**Pull request merged**
